test/cpp/api/dataloader.cpp

```diff
@@ -1274,3 +1275,111 @@ TEST(DataLoaderTest, StatefulDatasetWithCollate) {
   ASSERT_TRUE(batch->data[0].allclose(torch::ones(kBatchSize + 1)));
   ASSERT_TRUE(batch->target[0].allclose(torch::zeros(kBatchSize - 1)));
 }
+
+class DummyChunkDataSet : public datasets::ChunkDataSet<
+                         DummyChunkDataSet,
+                         std::vector<int>,
+                         samplers::SequentialSampler,
+                         samplers::SequentialSampler> {
+ public:
+  using BatchType = torch::optional<std::vector<int>>;
+  using BatchRequestType = size_t;
+  DummyChunkDataSet(size_t num_chunks, size_t batch_size)
+      : datasets::ChunkDataSet<
+            DummyChunkDataSet,
+            std::vector<int>,
+            samplers::SequentialSampler,
+            samplers::SequentialSampler>(),
+        num_chunks_(num_chunks),
+        batch_size_(batch_size),
+        chunk_sampler_(std::move(samplers::SequentialSampler(num_chunks))),
+        example_sampler_(std::move(samplers::SequentialSampler(batch_size))) {}
+
+  std::vector<int> read_chunk(size_t chunk_index) override {
+    std::vector<int> batch(batch_size_);
+    size_t counter = chunk_index * batch_size_;
+    for (auto& i : batch) {
+      i = counter++;
+    }
+    return batch;
+  }
+
+  /// Simply returns an entire chunk to test the API for now.
+  torch::optional<std::vector<int>> get_batch(size_t batch_size) override {
+    int index = chunk_index_.fetch_add(1);
+    if (index < num_chunks_) {
+      return read_chunk(index);
+    }
+    return torch::nullopt;
+  }
+
+  samplers::SequentialSampler get_chunk_sampler() override {
+    return chunk_sampler_;
+  }
+
+  samplers::SequentialSampler get_example_sampler() override {
+    return example_sampler_;
+  }
+
+  size_t get_chunk_count() override {
+    return num_chunks_;
+  }
+
+ private:
+  std::atomic<int> chunk_index_{0};
+  size_t num_chunks_;
+  size_t batch_size_;
+  samplers::SequentialSampler chunk_sampler_;
+  samplers::SequentialSampler example_sampler_;
+};
```

Comment:
I'm very confused by this example, because it seems that you always return a whole chunk as a batch, while if I understand correctly, the whole purpose of this dataset is to be able to load "large" chunks and produce multiple "small" batches from each one. Isn't that what we should be testing instead?

Reply:
@apaszke Thank you for looking at the PR. It is just the API I sent to get your comments. The implementation will do exactly as you mentioned, make smaller batches out of chunks and it will be a generic implementation so that different dataset types only need to implement read_chunk() methods. Since a full implementation is bit more complicated, we thought of sending the API first and seems like that make things bit more confusing :)